tree.go

```diff
@@ -1368,20 +1361,26 @@ func (n *InternalNode) serializeWithCompressedChildren(compressedChildren [][32]
 			setBit(bitlist, i)
 			if _, ok := c.(*HashedNode); ok {
 				setBit(hashlist[:], i)
-				nhashed++
+				hashedCount++
+			} else {
+				nonHashedCount++
 			}
 		}
 	}
 
-	ret = ret[:nodeTypeSize+bitlistSize+(len(compressedChildren)+nhashed)*SerializedPointCompressedSize]
+	ret = ret[:nodeTypeSize+bitlistSize+(nonHashedCount+hashedCount)*SerializedPointCompressedSize]
 	children := ret[internalNodeChildrenOffset:internalNodeChildrenOffset]
 	consumed := 0
 	for i := 0; i < NodeWidth; i++ {
 		if bit(bitlist, i) {
 			if bit(hashlist[:], i) {
 				children = append(children, n.children[i].(*HashedNode).commitment...)
 			} else {
-				children = append(children, compressedChildren[consumed][:]...)
+				childIdx, ok := compressedPointsIdxs[n.children[i]]
+				if !ok {
+					panic("children commitment not found in cache")
```

Comment:
This should never happen, but I'm not a fan of panics (which we already have in many other places).

Maybe it's worth, after things stabilize, to reconsider all panics and always return errors. In any case, the "error" is fatal anyway since go-ethereum won't be able to recover so not sure.